### DISCUSSION TITLE
MVUX and C# Markup getting fullscreen using a button

### DISCUSSION THREAD

**Discussion by emem67** [unoplatform/uno#18136] [Q&A] [answered]
I'm experimenting as in the subject and I managed to get in fullscreen in WASM and in a native windows application.
I have not found a way for Desktop application (Skia) in windows envinroment.
To get fullscreen in WASM I used this:

        // Works for WASM not for Desktop
        Windows.UI.ViewManagement.ApplicationView.GetForCurrentView().TryEnterFullScreenMode();

To do the same in native windows:

        // Works with WinApp not for Desktop  
        var app = Application.Current as App;  
        app.Presenter.Maximize();  

note I had to add a Presenter property in the App class
   
        public OverlappedPresenter Presenter { get; private set; }
        
        // Later in the OnLaunched
        // Set up for WinApp
        Presenter = (OverlappedPresenter)MainWindow.AppWindow.Presenter;

I created a small app to illustrate the concepts. 
[FullScreen.zip](https://github.com/user-attachments/files/16903631/FullScreen.zip)

I would like to do the same for the Desktop application in windows enviroment but I have not fond a way, any suggestion?
Thanks

**Comment by mcNets** [comment]
I'm using this code to maximize a desktop app in my linux device:
```csharp
if (Environment.OSVersion.Platform == PlatformID.Unix) {
    ApplicationView.GetForCurrentView().TryEnterFullScreenMode();
}
```

**Reply by emem67** [reply]
Thanks for your answer, this is what I did in WASM but it does not work in Desktop application, my guess is you are using Desktop in Linux whist I tried Desktop in Windows envinroment. (I will edit my post to be more clear)
Probably using desktop in windows envinroment is not to wise but in theory it is possible.

**Reply by mcNets** [reply]
I've tried in WIndows and it works for me.

**Reply by emem67** [reply]
Uh, then I do not know what to think.
As you can see here I have 'System.InvalidOperationException' in WindowsBase.dll (Desktop on Windows)

![image](https://github.com/user-attachments/assets/05517054-b6b9-4d2c-aad7-4f872f280176)

Maybe the issue is I'm tring to resize the main window and I have not created a second windows and the nresize it.
thanks anyway

**Reply by mcNets** [reply]
I'm maximazing it just after the activation.

**Reply by emem67** [reply]
I see, if I do this in the App.xaml.cs after the MainWindow.Activate(); it works.
If I do this from my page it doesent even if I create a property in the aa class like this

`public ApplicationView AppView { get; private set; }`

then this code

```
        
        // Ensure the current window is active
        MainWindow.Activate();

        AppView = ApplicationView.GetForCurrentView();

```

if in my page I do this I still have the problem

```      
        var app = Application.Current as App;
        app.AppView.TryEnterFullScreenMode();
```

**Reply by mcNets** [reply]
Are you getting a log error msg?

https://github.com/unoplatform/uno/blob/master/src/Uno.UI.Runtime.Skia.Gtk/Extensions/GtkApplicationViewExtension.cs

**Reply by emem67** [reply]
Sorry I'm not sure on where to look for logs, in the bin/debug folder there is anything.
Could you try my application? It is the zip file I uploaded earlier.

**Comment by mcNets** [comment, marked as answer]
@emem67  Try this:

1- Add a `Maximize()` method to the `Application` class: (I would move the logic for all TFM here.)
```csharp
public void MaximizeWindow()
{
    var appWindow = MainWindow?.AppWindow ?? throw new InvalidOperationException("AppWindow is null");

    MainWindow?.DispatcherQueue.TryEnqueue(() =>
    {
        if (appWindow.Presenter is FullScreenPresenter)
        {
            appWindow.SetPresenter(AppWindowPresenterKind.Overlapped);
        }
        else
        {
            appWindow.SetPresenter(AppWindowPresenterKind.FullScreen);
        }
    });
}
```

2- Change the ViewModel
```csharp
#if HAS_UNO_SKIA
        var app = Application.Current as App;
        app?.Maximize();
```

**Reply by emem67** [reply]
Wow you are on another legue :-) 
What you wrote works not only for Skia desktop but for WASM too!
ApplicationView.GetForCurrentView() still fails for native windows so I added my previous code to the Maximize method.
I tried to understand the concept of sending messges to the DispatcherQueue from anothet thread but I got lost.
I suspect the DispatcherQueue can work also for native windows but it is too much for me.
So your function become like this:

```
#if HAS_UNO
        var appWindow = MainWindow?.AppWindow ?? throw new InvalidOperationException("AppWindow is null");
        MainWindow?.DispatcherQueue.TryEnqueue(() =>
        {
            if (appWindow.Presenter is FullScreenPresenter)
            {
                appWindow.SetPresenter(AppWindowPresenterKind.Overlapped);
            }
            else
            {
                appWindow.SetPresenter(AppWindowPresenterKind.FullScreen);
            }
        });
#else
        if(Presenter.State == OverlappedPresenterState.Maximized)
        {
            Presenter.Restore();
        }
        else
        {
            Presenter.Maximize();
        }
#endif
```

off course I have to create a variable for the presenter in the App class

`private OverlappedPresenter Presenter { get; set; }`

and set it at the end of the OnLaunched method

`Presenter = (OverlappedPresenter)MainWindow.AppWindow.Presenter;`

BTW I consider your suggertion the answer to this thread
many many thanks!

**Reply by mcNets** [reply]
You can get the `Presenter` within the Maximize function too, like appView

**Reply by MartinZikmund** [reply]
@emem67 @mcNets I notice you are using `ApplicationView` in case of Uno, however `AppWindow` presenters should now be supported on Uno as well, and actually should be more reliable than using `ApplicationView` which is officialy deprecated in WinUI - so try using `FullScreenPresenter` for the `AppWindow` if possible

**Reply by mcNets** [reply]
@MartinZikmund changed.

**Reply by emem67** [reply]
I also update my code snippet that includes the native windows part, thanks!
